pkg/collector/smi_collector.go

```diff
@@ -0,0 +1,109 @@
+package collector
+
+import (
+	"log"
+	"path/filepath"
+	"strings"
+
+	"github.com/Azure/aks-periscope/pkg/interfaces"
+	"github.com/Azure/aks-periscope/pkg/utils"
+)
+
+// SmiCollector defines an Smi Collector struct
+type SmiCollector struct {
+	BaseCollector
+}
+
+var _ interfaces.Collector = &SmiCollector{}
+
+// NewSmiCollector is a constructor
+func NewSmiCollector(exporter interfaces.Exporter) *SmiCollector {
+	return &SmiCollector{
+		BaseCollector: BaseCollector{
+			collectorType: Smi,
+			exporter:      exporter,
+		},
+	}
+}
+
+// Collect implements the interface method
+func (collector *SmiCollector) Collect() error {
+	// Get all CustomResourceDefinitions in the cluster
+	allCrdsList, err := utils.GetResourceList([]string{"get", "crds", "-o", "jsonpath={..metadata.name}"}, " ")
+	if err != nil {
+		return err
+	}
+
+	// Directory where logs will be written to
+	rootPath, err := utils.CreateCollectorDir(collector.GetName())
+	if err != nil {
+		return err
+	}
+
+	// Filter to obtain a list of Smi CustomResourceDefinitions in the cluster
+	crdNameContainsSmiPredicate := func(s string) bool { return strings.Contains(s, "smi-spec.io") }
+	smiCrdsList := filter(allCrdsList, crdNameContainsSmiPredicate)
+	if len(smiCrdsList) == 0 {
+		// TODO should this return an error?
+		log.Printf("Cluster does not contain any SMI CustomResourceDefinitions\n")
+		return nil
+	}
+
+	collectSmiCrdDefinitions(collector, filepath.Join(rootPath, "smi_crd_definitions"), smiCrdsList)
+	collectSmiCustomResourcesFromAllNamespaces(collector, filepath.Join(rootPath, "smi_custom_resources"), smiCrdsList)
+
+	return nil
+}
+
+func collectSmiCrdDefinitions(collector *SmiCollector, rootPath string, smiCrdsList []string) {
+	for _, smiCrd := range smiCrdsList {
+		fileName := smiCrd + "_definition.yaml"
+		kubeCmd := []string{"get", "crd", smiCrd, "-o", "yaml"}
+		if err := collector.CollectKubectlOutputToCollectorFiles(rootPath, fileName, kubeCmd); err != nil {
+			log.Printf("Failed to collect %s: %+v", fileName, err)
+		}
+	}
+}
+
+func collectSmiCustomResourcesFromAllNamespaces(collector *SmiCollector, rootPath string, smiCrdsList []string) {
+	// Get all namespaces in the cluster
+	namespacesList, err := utils.GetResourceList([]string{"get", "namespaces", "-o", "jsonpath={..metadata.name}"}, " ")
+	if err != nil {
+		log.Printf("Failed to list namespaces in the cluster: %+v", err)
```

Comment:
💡 same as above: Worth identifying that use-case here is `error` or something else?

Reply:
Resolved in 5b56484.
This should return an error. Periscope needs to list all of the cluster's namespaces in order to proceed with collecting SMI custom resources. If Periscope fails to list the cluster's namespaces (`kubectl get namespaces...`), then something horrible has gone wrong in the cluster, so an error should be returned here.

```
	// Get all namespaces in the cluster
	namespacesList, err := utils.GetResourceList([]string{"get", "namespaces", "-o", "jsonpath={..metadata.name}"}, " ")
	if err != nil {
		return fmt.Errorf("Failed to collect SMI custom resources: unable to list namespaces in the cluster: %+v", err)
	}
```